src/commonMain/kotlin/fr/acinq/bitcoin/Crypto.kt

```diff
@@ -154,9 +154,32 @@ public object Crypto {
         )
     }
 
+    /**
+     * @param data data to sign (32 bytes)
+     * @param privateKey private key
+     * @param merkleRoot optional tapscript merkle root, which will be used to tweak the private key.
+     * if merkleRoot is null, no tweaking is done, sign with key directly (this is used for signatures in BIP342 script)
+     * if merkleRoot is all zeroes, sign with privateKey + H_TapTweak(public key) (this is used for key path spending when no scripts are present).
+     * Otherwise, sign with key + H_TapTweak(public key || merkle_root) (this is used for key path spending, with specific Merkle root of the script tree).
+     * @param auxrand32 optional auxiliary random data
+     * @return the Schnorr signature of data with private key (optionally tweaked with the tapscript merkle root)
+     */
+    @JvmStatic
+    public fun signSchnorr(data: ByteVector32, privateKey: PrivateKey, merkleRoot: ByteVector32?, auxrand32: ByteVector32? = null): ByteVector64 {
```

Comment:
As discussed offline, it would be great to use a sealed class instead of the merkleRoot here, something like:

```kotlin
sealed class SchnorrTweak {
    /** Use this when producing signatures inside BIP342 scripts. */
    object NoTweak : SchnorrTweak()
    /** Use this for P2TR that don't have a script tree. */
    object TweakNoScript : SchnorrTweak()
    /** Use this for P2TR that contain a script tree. */
    data class Tweak(val merkleRoot: ByteVector32) : SchnorrTweak()
}
```

```suggestion
    public fun signSchnorr(data: ByteVector32, privateKey: PrivateKey, tweak: SchnorrTweak, auxrand32: ByteVector32? = null): ByteVector64 {
```

Reply:
I agree this  better clarifies what's happening. From a code readability standpoint I would even go further to define different objects for generic Schnorr tweaks and TR specific tweaks. This would result in code that looks something like:

```kotlin
// using explicit Taproot tweaks
sig = signSchnorr(hash, scriptPrivKey, ScriptPathTweak)
sig = signSchnorr(hash, internalPrivKey, NoScriptPathsTweak)
sig = signSchnorr(hash, internalPrivKey, KeyPathTweak(merkleRoot))

// using generic Schnorr tweaks
sig = sigSchnorr(hash, privKey, NoTweak)
sig = sigSchnorr(hash, privKey, Tweak(byteVector32)

```